File: test-suite/src/test/java/io/micronaut/kafka/docs/consumer/scope/ProductListener.java
package io.micronaut.kafka.docs.consumer.scope;

// tag::imports[]
import io.micronaut.configuration.kafka.annotation.KafkaListener;
import io.micronaut.configuration.kafka.annotation.KafkaScope;
import io.micronaut.configuration.kafka.annotation.OffsetReset;
import io.micronaut.configuration.kafka.annotation.Topic;
import jakarta.inject.Inject;
// end::imports[]

// tag::listener[]
@KafkaListener(offsetReset = OffsetReset.EARLIEST)
public class ProductListener {

    // tag::scope[]
    @KafkaScope
    static class ProductMetadata {
        private final String correlationId = java.util.UUID.randomUUID().toString();

        String getCorrelationId() {
            return correlationId;
        }
    }
    // end::scope[]

    @Inject ProductMetadata productMetadata;

    @Topic("products")
    void receive(String product) {
        String correlationId = productMetadata.getCorrelationId();
        // use the same metadata instance throughout this listener invocation
    }
}
// end::listener[]

File: test-suite-kotlin/src/test/kotlin/io/micronaut/kafka/docs/consumer/scope/ProductListener.kt
package io.micronaut.kafka.docs.consumer.scope

import io.micronaut.configuration.kafka.annotation.KafkaListener
import io.micronaut.configuration.kafka.annotation.KafkaScope
import io.micronaut.configuration.kafka.annotation.OffsetReset
import io.micronaut.configuration.kafka.annotation.Topic
import jakarta.inject.Inject
import java.util.UUID

// tag::listener[]
@KafkaListener(offsetReset = OffsetReset.EARLIEST)
class ProductListener {

    // tag::scope[]
    @KafkaScope
    open class ProductMetadata {
        private val correlationId: String = UUID.randomUUID().toString()

        open fun correlationId(): String = correlationId
    }
    // end::scope[]

    @Inject lateinit var productMetadata: ProductMetadata

    @Topic("products")
    fun receive(product: String) {
        println("Received $product with correlation ${productMetadata.correlationId()}")
    }
}
// end::listener[]
